File: backend2.0/src/domain/entities/vision_usage.py
"""Vision API usage metadata — real token counts from provider responses."""
from __future__ import annotations

from dataclasses import dataclass, field

from providers.vision.base import TokenUsage


def _token_dict(usage: TokenUsage | None) -> dict[str, int | None]:
    if usage is None:
        return {
            "prompt_token_count": None,
            "candidates_token_count": None,
            "thoughts_token_count": None,
            "total_token_count": None,
        }
    return {
        "prompt_token_count": usage.prompt_token_count,
        "candidates_token_count": usage.candidates_token_count,
        "thoughts_token_count": usage.thoughts_token_count,
        "total_token_count": usage.total_token_count,
    }


@dataclass
class VisionPassRecord:
    """One vision API call (extraction or correction pass)."""

    pass_number: int
    pass_kind: str  # "extraction" | "correction"
    provider: str
    model: str | None
    usage: TokenUsage
    page_number: int | None = None
    correction_mode: str | None = None  # "selective" | "full_fallback"
    correction_fields: dict | None = None

    def to_dict(self) -> dict:
        payload = {
            "pass": self.pass_number,
            "pass_kind": self.pass_kind,
            "provider": self.provider,
            "model": self.model,
            "page_number": self.page_number,
            **_token_dict(self.usage),
        }
        if self.correction_mode:
            payload["correction_mode"] = self.correction_mode
        if self.correction_fields is not None:
            payload["correction_fields"] = self.correction_fields
        return payload


@dataclass
class VisionUsageSnapshot:
    """Aggregated usage for one page/region (1–2 API calls)."""

    passes: list[VisionPassRecord] = field(default_factory=list)
    totals: TokenUsage = field(default_factory=TokenUsage)
    api_calls: int = 0
    page_number: int | None = None

    def merge(self, other: VisionUsageSnapshot | None) -> VisionUsageSnapshot:
        if other is None:
            return self
        return VisionUsageSnapshot(
            passes=[*self.passes, *other.passes],
            totals=self.totals.add(other.totals),
            api_calls=self.api_calls + other.api_calls,
            page_number=self.page_number or other.page_number,
        )

    def to_dict(self) -> dict:
        return {
            "page_number": self.page_number,
            "api_calls": self.api_calls,
            "passes": [p.to_dict() for p in self.passes],
            "totals": _token_dict(self.totals),
        }


@dataclass
class JobVisionUsage:
    """Job-level vision usage across all pages/regions."""

    pages: list[VisionUsageSnapshot] = field(default_factory=list)
    totals: TokenUsage = field(default_factory=TokenUsage)
    api_calls: int = 0

    @classmethod
    def from_page_snapshots(cls, snapshots: list[VisionUsageSnapshot | None]) -> JobVisionUsage | None:
        valid = [s for s in snapshots if s is not None and s.api_calls > 0]
        if not valid:
            return None

        totals = TokenUsage()
        api_calls = 0
        for snap in valid:
            totals = totals.add(snap.totals)
            api_calls += snap.api_calls

        return cls(pages=valid, totals=totals, api_calls=api_calls)

    @property
    def extraction_model(self) -> str | None:
        for page in self.pages:
            for record in page.passes:
                if record.pass_kind == "extraction" and record.model:
                    return record.model
        return None

    @property
    def correction_model(self) -> str | None:
        for page in self.pages:
            for record in page.passes:
                if record.pass_kind == "correction" and record.model:
                    return record.model
        return None

    @property
    def extraction_provider(self) -> str | None:
        for page in self.pages:
            for record in page.passes:
                if record.pass_kind == "extraction":
                    return record.provider
        return None

    @property
    def correction_provider(self) -> str | None:
        for page in self.pages:
            for record in page.passes:
                if record.pass_kind == "correction":
                    return record.provider
        return None

    def all_passes(self) -> list[VisionPassRecord]:
        out: list[VisionPassRecord] = []
        for page in self.pages:
            out.extend(page.passes)
        return out

    def to_dict(self) -> dict:
        passes = [p.to_dict() for p in self.all_passes()]
        models = {
            "extraction": {
                "provider": self.extraction_provider,
                "model": self.extraction_model,
            },
            "correction": {
                "provider": self.correction_provider,
                "model": self.correction_model,
            },
        }
        return {
            "api_calls": self.api_calls,
            "passes": passes,
            "pages": [p.to_dict() for p in self.pages],
            "totals": _token_dict(self.totals),
            "models": models,
        }

File: frontend/src/utils/testSuite/visionUsage.js
/**
 * Vision API usage helpers for the test-suite.
 * All token counts come from backend provider responses — never estimated client-side.
 */

/** @typedef {{ prompt_token_count?: number|null, candidates_token_count?: number|null, thoughts_token_count?: number|null, total_token_count?: number|null }} TokenTotals */

/** @typedef {{ pass?: number, pass_kind?: string, provider?: string, model?: string|null, page_number?: number|null, correction_mode?: string|null, correction_fields?: object|null, prompt_token_count?: number|null, candidates_token_count?: number|null, thoughts_token_count?: number|null, total_token_count?: number|null }} VisionPass */

/** @typedef {{ api_calls?: number, passes?: VisionPass[], pages?: object[], totals?: TokenTotals, models?: { extraction?: { provider?: string|null, model?: string|null }, correction?: { provider?: string|null, model?: string|null } } }} VisionUsage */

export function formatTokenCount(value) {
  if (value == null || !Number.isFinite(value)) return '—'
  return value.toLocaleString()
}

export function getVisionUsage(aiResult) {
  return aiResult?.vision_usage ?? null
}

export function hasVisionUsage(aiResult) {
  const usage = getVisionUsage(aiResult)
  return Boolean(usage && (usage.api_calls > 0 || usage.totals?.total_token_count != null))
}

/** Sum token fields across usage objects (real API values only). */
export function sumTokenTotals(usages) {
  /** @type {TokenTotals} */
  const out = {
    prompt_token_count: null,
    candidates_token_count: null,
    thoughts_token_count: null,
    total_token_count: null,
  }

  for (const usage of usages) {
    const totals = usage?.totals
    if (!totals) continue
    for (const key of Object.keys(out)) {
      const val = totals[key]
      if (val == null || !Number.isFinite(val)) continue
      out[key] = (out[key] ?? 0) + val
    }
  }

  return out
}

export function aggregateBatchVisionUsage(results) {
  const usages = results
    .map((row) => getVisionUsage(row.aiResult))
    .filter(Boolean)

  if (!usages.length) return null

  const totals = sumTokenTotals(usages)
  const apiCalls = usages.reduce((sum, u) => sum + (u.api_calls ?? 0), 0)

  const extractionModels = new Set()
  const correctionModels = new Set()
  for (const usage of usages) {
    const ext = usage.models?.extraction
    const corr = usage.models?.correction
    if (ext?.model) extractionModels.add(`${ext.provider ?? '?'}:${ext.model}`)
    if (corr?.model) correctionModels.add(`${corr.provider ?? '?'}:${corr.model}`)
  }

  return {
    caseCount: usages.length,
    apiCalls,
    totals,
    extractionModels: [...extractionModels],
    correctionModels: [...correctionModels],
  }
}

export function passLabel(pass) {
  if (!pass) return '—'
  const num = pass.pass ?? pass.pass_number
  const kind = pass.pass_kind === 'correction' ? 'Pass 2 · correction' : 'Pass 1 · extraction'
  return num ? `${kind} (#${num})` : kind
}

export function formatModelRef(ref) {
  if (!ref?.model && !ref?.provider) return '—'
  if (ref.model && ref.provider) return `${ref.provider} / ${ref.model}`
  return ref.model ?? ref.provider ?? '—'
}

export function correctionFieldsSummary(fields) {
  if (!fields || typeof fields !== 'object') return null
  if (fields.mode === 'full_fallback') {
    return `Full fallback (pass-1 raw text, ${fields.pass1_text_length ?? '?'} chars)`
  }
  const parts = []
  if (fields.scan_for_missed_rooms) parts.push('scan missed rooms')
  const verifyCount = Array.isArray(fields.rooms_to_verify) ? fields.rooms_to_verify.length : 0
  if (verifyCount) parts.push(`${verifyCount} room(s) to verify`)
  if (fields.floor_identification) parts.push('floor identification')
  if (Array.isArray(fields.existing_room_names) && fields.existing_room_names.length) {
    parts.push(`${fields.existing_room_names.length} existing names`)
  }
  return parts.length ? parts.join(' · ') : 'Selective correction payload'
}

export function tokenBreakdownLine(totals) {
  if (!totals) return '—'
  const prompt = formatTokenCount(totals.prompt_token_count)
  const output = formatTokenCount(totals.candidates_token_count)
  const thoughts = formatTokenCount(totals.thoughts_token_count)
  const total = formatTokenCount(totals.total_token_count)
  const parts = [`prompt ${prompt}`, `output ${output}`]
  if (totals.thoughts_token_count != null) parts.push(`thoughts ${thoughts}`)
  parts.push(`total ${total}`)
  return parts.join(' · ')
}
